### PR TITLE
ARMOCP-417: enable arm64 for agent installer

### PR DESCRIPTION
**Pull request by jeffdyoung**
Translate arm64-> aarch64 in instances where we reference rhcos (should mirror x86_64 -> amd64 mapping)
In parallel with openshift-installer change for agent installer: https://github.com/openshift/installer/pull/6401
## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment (dev-scripts needs additional work to support arm)
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
Manually created agent.iso and installed OCP with arm64 BM HW in our RDU2 lab
- [] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/bminventory/inventory.go
Comment (cs-zhang) [inline internal/bminventory/inventory.go:3960]: Should not fix the arch to x86 and arm, the original check should work for multi-arch support.

Comment (jeffdyoung) [reply internal/bminventory/inventory.go]: @cs-zhang I moved this logic out of the MultiCPUArchitecture block. It needs to be there otherwise I get a `release = aarch64 and cpu arch = arm64 does` not match error

File: internal/versions/versions.go
Comment (mkowalski) [inline internal/versions/versions.go:143]: If we take this path, can we use both of those as `common.*`? E.g. `common.ARM64CPUArchitecture` and `common.AARCH64CPUArchitecture` or similar

File: internal/versions/versions.go
Comment (mkowalski) [inline internal/versions/versions.go:237]: This function has quite an extensive set of unit tests in around https://github.com/openshift/assisted-service/blob/master/internal/versions/versions_test.go#L929-L934 and https://github.com/openshift/assisted-service/blob/master/internal/versions/versions_test.go#L403-L421

It would be very appreciated if you created a set of cases there that cover `GetOsImage` via `arm64` as well as via `aarch64` so that by reading the test I can understand the rationale behind the change.

Look e.g. at the test https://github.com/openshift/assisted-service/blob/master/internal/versions/versions_test.go#L579-L592 where we explicitly have a statement "gets X using X query" and "gets X using Y query"

File: internal/bminventory/inventory.go
Comment (mkowalski) [inline internal/bminventory/inventory.go:1540]: Instead of modifying the content of this conditional, could we maybe modify the content of the function `isBaremetalBinaryFromAnotherReleaseImageRequired` so that it does not enter this code at all? AFAIK this block is meant to support cross-installation where Hub is x86 but spoke is ARM (but not the reverse). I.e. this block handles a scenario when `openshift-install` is executed on platform x86 but manifests that it generate may be used by ARM. By writing here `cluster.CPUArchitecture` you are saying "use openshift-install binary that matches the installed cluster architecture" which is different from "use openshift-install binary that matches the Hub cluster architecture"

Comment (mkowalski) [reply internal/bminventory/inventory.go:1540]: TLDR: Have you tested that with this change you can install single-arch ARM cluster from X86 hub with no regression?

File: internal/controller/controllers/infraenv_controller.go
Comment (mkowalski) [inline internal/controller/controllers/infraenv_controller.go:348]: Maybe too quick and missed some other logic, but don't you want to write here `... := NormalizeRHCOSArch(infraEnv.Spec.CpuArchitecture)`?

Comment (jeffdyoung) [reply internal/controller/controllers/infraenv_controller.go:348]: I meant to remove this all together, but will add `NormalizeRHCOSArch(infraEnv.Spec.CpuArchitecture)` back based on our slack conversation.

File: api/v1beta1/agentserviceconfig_types.go
Comment (jkilzi) [inline api/v1beta1/agentserviceconfig_types.go:39]: Can we make this an enum instead of any string?

**Comment by jeffdyoung** [comment]
Hey @mkowalski @jkilzi Do you guys have time to re-review? Its been a few months, so to refresh memories:

These changes are needed b/c the AI expects `arm64` for its cpu_architecture values. The agent installer will produce `aarch64` artifacts, because it generates os_images and release_image straight from the release image (like `x86_64`)  This PR's purpose is to allow agent to use `aarch64` without breaking the other AI components that require `arm64`.

After this merges, revisit migrating the entire assisted stack to use `aarch64`: https://issues.redhat.com/browse/ARMOCP-487

Arm Agent installer PR: https://github.com/openshift/installer/pull/6401

File: swagger.yaml
Comment (carbonin) [inline swagger.yaml:4394]: What purpose does the empty value in this enum serve if there's also a default set?
(applies for all instances of this in the swagger file).

Comment (jeffdyoung) [reply swagger.yaml:4394]: Ack, missed that... Thanks.

**Comment by carbonin** [comment]
Do we think we need to normalize [these](https://github.com/openshift/assisted-service/blob/204d6fc0d4601690f41cb5c38ef3ced4fb00b17b/internal/controller/controllers/preprovisioningimage_controller.go#L131)?

This is comparing the architecture in the infraEnv CR with the one in the preprovisioningimage (detected by ironic in some way).

**Comment by jeffdyoung** [comment]
> Do we think we need to normalize [these](https://github.com/openshift/assisted-service/blob/204d6fc0d4601690f41cb5c38ef3ced4fb00b17b/internal/controller/controllers/preprovisioningimage_controller.go#L131)?
> 
I haven't hit this in my local testing w/agent installer. FWIW, my first attempts was normalizing all comparisons, didn't work and was kinda messy. Normalizing the inputs seems to be the best route. 

> This is comparing the architecture in the infraEnv CR with the one in the preprovisioningimage (detected by ironic in some way).
> 
> Although I don't think this PR will change either of those as one is provided by the user and the other is provided by ironic so maybe it's fine (or at least as fine as it was before).

**Comment by carbonin** [comment]
>I haven't hit this in my local testing w/agent installer.

You wouldn't, this is a ZTP-specific section of the code, but I think we can leave it for now.

Looks like you'll need to address the subsystem test failure though.

File: subsystem/cluster_v2_test.go
Comment (carbonin) [inline subsystem/cluster_v2_test.go:492]: Ah I guess this is failing because you added the enum, right?

Can you also just change the name of the test?
Maybe something like `Fails to register an infraenv with a non-supported OS architecture` or something to distinguish it from the release/OS images provided in the config since that's actually not relevant to this test anymore.

File: swagger.yaml
Comment (carbonin) [inline swagger.yaml:458]: Not sure I understand this change. When would we use this?

Comment (carbonin) [reply swagger.yaml:458]: Yeah, subsystem tests are kinda annoying to run locally, but you should be able to just get the system up and running in minikube with `make deploy-test` then just run the query the subsystem test is running manually.

Comment (carbonin) [reply swagger.yaml:458]: Also, however you want to handle testing is fine, I just wasn't sure if I should review this or if you were just testing, maybe mark the whole PR as WIP until you want reviews on the changes?

**Review by carbonin — APPROVED** [review]

**Comment by carbonin** [comment]
FWIW this also resolves https://issues.redhat.com/browse/MGMT-13386

**Pull request merged**
